### PR TITLE
[dist] feat:  No resharding enabled for accelerated small model training

### PR DESCRIPTION
**Pull request by yangtian6781**
### What does this PR do?

Add switches to control reshard_after_forward and reshard_after_backward, enabling faster training for small models under gradient accumulation.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `misc`, `ci`, `config`, `docs`, `data`, `dist`, `omni`, `logging`, `model`, `optim`, `ckpt`, `release`, `task`, `perf`, `ops`, `parallel`
  - If this PR involves multiple modules, separate them with `,` like `[ci, data, model]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][parallel, model] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md?plain=1#L22): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/ByteDance-Seed/VeOmni/blob/main/docs).
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/ByteDance-Seed/VeOmni/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...

### CODE REVIEW

File: tasks/train_torch.py [hunk 1]
```diff
@@ -274,8 +275,18 @@ def main():
...
+            for micro_step, micro_batch in enumerate(micro_batches):
+                if (
+                    args.train.data_parallel_mode == "fsdp2"
+                    and not args.train.enable_reshard_after_backward
+                    and num_micro_steps > 0
+                ):
+                    if micro_step == 0:
+                        model.set_reshard_after_backward(False)
+                    elif micro_step == num_micro_steps - 1:
+                        model.set_reshard_after_backward(True)
```
[the remark this answers is not in the document]
Comment (yangtian6781) [reply tasks/train_torch.py:290]: Oh, there is indeed a bug here.

File: tasks/omni/train_flux.py [hunk 2]
```diff
@@ -429,7 +430,18 @@ def main():
...
+            for micro_step, batch in enumerate(micro_batches):
+                if (
+                    args.train.data_parallel_mode == "fsdp2"
+                    and not args.train.enable_reshard_after_backward
+                    and num_micro_steps > 1
+                ):
+                    if micro_step == 0:
+                        model.set_reshard_after_backward(False)
+                    elif micro_step == num_micro_steps - 1:
+                        model.set_reshard_after_backward(True)
```
Comment (Luosuu) [inline tasks/omni/train_flux.py:444]: do we need to set this to recursive?

File: veomni/arguments/arguments_types.py [hunk 3]
```diff
@@ -400,6 +400,14 @@ class TrainingArguments:
         default=1.0,
         metadata={"help": "Ratio of learning rate decay steps."},
     )
+    enable_reshard_after_forward: bool = field(
+        default=True,
+        metadata={"help": "Enabl reshard after forward for FSDP2."},
```
Comment (Luosuu) [inline veomni/arguments/arguments_types.py]: fix typo

**Review by Luosuu — COMMENTED** [review]
please also give a short experiment report to show
* whether grad_norm/loss is impacted
* how much training speed changes for small models

**Comment by yangtian6781** [comment]
> please also give a short experiment report to show
> 
> * whether grad_norm/loss is impacted
> * how much training speed changes for small models

My experimental setup is as follows:
Computing power: 4 x a100
model.model_path Qwen/Qwen3-1.7B
data.train_path VeOmni/tulu-first2000.parquet
train.global_batch_size 128
train.micro_batch_size 1
train.dyn_bsz false
train.num_train_epochs 1
train.enable_full_determinism true

bitwise check experiment:
image
The no_shard setting achieves a higher MFU：
image

File: tasks/omni/train_flux.py (hunk 2 above)
Comment (yangtian6781) [reply tasks/omni/train_flux.py:444]: https://github.com/pytorch/pytorch/blob/0f84569b23274622270f0a8186e7ad82131be4e7/torch/distributed/fsdp/_fully_shard/_fully_shard.py#L410
We only need to set it at the outermost level of the FSDPModule.

File: veomni/arguments/arguments_types.py (hunk 3 above)
Comment (yangtian6781) [reply veomni/arguments/arguments_types.py]: done

**Comment by Luosuu** [comment]
Sorry our CI is currently down. Let me find someone to fix it and then I will merge it.

**Comment by Luosuu** [comment]
Hi @yangtian6781 ,  sorry but please resolve the conflict and then I will merge it.
